events/label_events.go

```diff
@@ -0,0 +1,151 @@
+package events
+
+import (
+	"fmt"
+)
+
+type LabelEventManager struct {
+	subs []*LabelSubscriber
+
+	ops        chan *LabelOperation
+	closed     chan struct{}
+	bufferSize int
+
+	persister LabelEventPersistence
+}
+
+func NewLabelEventManager() *LabelEventManager {
+	return &LabelEventManager{
+		ops:        make(chan *LabelOperation),
+		closed:     make(chan struct{}),
+		bufferSize: 1024,
+		persister:  NewMemLabelPersister(),
+	}
+}
+
+type LabelOperation struct {
+	op  int
+	sub *LabelSubscriber
+	evt *LabelEvent
+}
+
+func (em *LabelEventManager) Run() {
+	for op := range em.ops {
+		switch op.op {
+		case opSubscribe:
+			em.subs = append(em.subs, op.sub)
+			op.sub.outgoing <- &LabelEvent{}
+		case opUnsubscribe:
+			for i, s := range em.subs {
+				if s == op.sub {
+					em.subs[i] = em.subs[len(em.subs)-1]
+					em.subs = em.subs[:len(em.subs)-1]
+					break
+				}
+			}
+		case opSend:
+			em.persister.Persist(op.evt)
+
+			for _, s := range em.subs {
+				if s.filter(op.evt) {
+					select {
+					case s.outgoing <- op.evt:
+					default:
+						log.Error("event overflow")
+					}
+				}
+			}
+		default:
+			log.Errorf("unrecognized eventmgr operation: %d", op.op)
+		}
+	}
+}
+
+type LabelSubscriber struct {
+	outgoing chan *LabelEvent
+
+	filter func(*LabelEvent) bool
+
+	done chan struct{}
+}
+
+type LabelEvent struct {
+	Seq    int64   `cborgen:"seq"`
+	Labels []Label `cborgen:"labels"`
+}
+
+// this is here, instead of under 'labeling' package, to avoid an import loop
+type Label struct {
+	LexiconTypeID string  `json:"$type" cborgen:"$type,const=app.bsky.label.label"`
+	SourceDid     string  `json:"src" cborgen:"src"`
+	SubjectUri    string  `json:"uri" cborgen:"uri"`
+	SubjectCid    *string `json:"cid,omitempty" cborgen:"cid"`
+	Value         string  `json:"val" cborgen:"val"`
+	Timestamp     string  `json:"ts" cborgen:"ts"` // TODO: actual timestamp?
```

Comment:
Re: the TODO, hard question!  I think ideally we use lex's datetime type for this in order to cut with the grain of lexicon, and if it needs to be more compact for some reason then we address that through lex/encodings in the future.

Reply:
Yeah, the plan here is to use codegen'd type from `com.atproto.label.label`, and inherit the correct type on this field from there.

Cross-referencing to https://github.com/bluesky-social/atproto/pull/525, where there is discussion about `labelUri` field.